src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.cs

```diff
@@ -696,6 +696,21 @@ internal void ConfigureForJsonSerializer()
                     case JsonSerializerContext ctx when AppContextSwitchHelper.IsSourceGenReflectionFallbackEnabled:
                         // .NET 6 compatibility mode: enable fallback to reflection metadata for JsonSerializerContext
                         _effectiveJsonTypeInfoResolver = JsonTypeInfoResolver.Combine(ctx, defaultResolver);
+
+                        if (_cachingContext is { } cachingContext)
```

Comment:
With this PR, the `JsonSerializerOptions` cache becomes the source of truth even for the `JsonSerializerContext` serialization overloads. This means that the cache can be populated even before the `ConfigureForJsonSerializer` is called for the first time, which is why we now need to invalidate the cache once it happens.

Reply:
any perf concern with that?

Reply:
Worst case scenario it would mean that sourcegen metadata need to be recreated from scratch once in the lifetime of the cache. It only impacts users that have turned on the compatibility feature switch.